### PR TITLE
feat: Add FirecrawlWebSearch component

### PR DESCRIPTION
**Pull request by saakshigupta2002**
### Related Issues

- part of #2870

### Proposed Changes:

This PR adds a new `FirecrawlWebSearch` component to the Firecrawl integration, enabling web search queries using the [Firecrawl Search API](https://docs.firecrawl.dev/features/search). This follows the pattern established by Haystack's existing `SearchApiWebSearch` and `SerperDevWebSearch` components, as suggested in the PR review for #2859.

Many use cases require starting from a user search query rather than from a predefined URL list. While the existing `FirecrawlCrawler` handles crawling known URLs, this new component allows users to search the web and retrieve results as Haystack Documents — making it straightforward to build search-augmented pipelines.

**What's included:**

- **`FirecrawlWebSearch` component** at `haystack_integrations.components.websearch.firecrawl`
  - Follows the standard Haystack WebSearch interface: accepts a `query` string, returns `documents` and `links`
  - Supports both synchronous (`run()`) and asynchronous (`run_async()`) execution
  - Configurable `top_k` for result limiting and `search_params` for Firecrawl-specific options (time filters, location, scrape options, etc.)
  - Full serialization support via `to_dict()` / `from_dict()`
  - Graceful error handling with logging (returns empty results on failure)
  - Handles both search-only results (title/description/url) and scraped document results (full markdown content when `scrapeOptions` are provided)

- **Comprehensive test suite** (16 unit tests + 2 integration tests)
  - Tests for initialization, serialization round-trip, search execution, parameter overrides, `top_k` truncation, async execution, error handling, warm-up behavior, and edge cases (empty/null results)

- **Configuration updates**
  - Updated `pyproject.toml` with `"Web Search"` keyword and mypy target for the new module
  - Updated `pydoc/config_docusaurus.yml` to include the new component in API documentation generation

**Usage example:**

```python
from haystack_integrations.components.websearch.firecrawl import FirecrawlWebSearch
from haystack.utils import Secret

websearch = FirecrawlWebSearch(
    api_key=Secret.from_env_var("FIRECRAWL_API_KEY"),
    top_k=5,
)
websearch.warm_up()

result = websearch.run(query="What is Haystack by deepset?")
documents = result["documents"]  # List of Document objects with search result content
links = result["links"]          # List of URL strings
```

**Design decisions:**

- Uses the Firecrawl Python SDK (`firecrawl-py`) directly rather than raw HTTP requests, consistent with the existing `FirecrawlCrawler` implementation. No new dependencies needed since `firecrawl-py>=4.0.0` already includes the `search()` method.
- Placed under `components/websearch/firecrawl/` to follow Haystack's namespace convention for web search components, separate from the existing crawler under `components/fetchers/firecrawl/`.
- `top_k` maps to Firecrawl's `limit` parameter when not explicitly set in `search_params`, providing a familiar interface while allowing full control through `search_params`.
- Response parsing handles both response types: `SearchResultWeb` objects (search metadata only) and Firecrawl `Document` objects (full markdown content when `scrapeOptions` are provided), using attribute detection rather than type imports to avoid coupling to SDK internals.

### How did you test it?

- All 16 unit tests pass (`hatch run test:unit`) — covering initialization, serialization round-trip, search execution with mocked clients, runtime parameter overrides, `top_k` truncation, async execution, error handling, warm-up behavior, and edge cases (empty/null web results)
- All existing `FirecrawlCrawler` tests continue to pass (14/14)
- Lint and format checks pass (`hatch run fmt-check`)
- Integration tests included (require `FIRECRAWL_API_KEY` environment variable): `test_run_integration` and `test_run_async_integration`

### Notes for the reviewer

- The new component lives under `components/websearch/firecrawl/` (separate from the existing `components/fetchers/firecrawl/`) to follow the same namespace convention as Haystack core's `SearchApiWebSearch` and `SerperDevWebSearch`
- The `_parse_search_response` method handles two types of results from the Firecrawl SDK: basic `SearchResultWeb` objects (when no scrape options are set) and full `Document` objects with markdown content (when `scrapeOptions` are provided). This is handled via `hasattr` checks rather than importing SDK-internal types.
- No new dependencies were added — `firecrawl-py>=4.0.0` (already a dependency) includes the `search()` method

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+    ```python
+    from haystack_integrations.components.websearch.firecrawl import FirecrawlWebSearch
+    from haystack.utils import Secret
+
+    websearch = FirecrawlWebSearch(
+        api_key=Secret.from_env_var("FIRECRAWL_API_KEY"),
+        top_k=5,
+    )
+    websearch.warm_up()
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]: `warm_up` is not needed as the component is automatically warmed up on the first call of `run`.
```suggestion
```

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            self,
+            api_key=self.api_key.to_dict(),
+            top_k=self.top_k,
+            search_params=self.search_params,
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict[str, Any]) -> "FirecrawlWebSearch":
+        """Deserializes the component from a dictionary."""
+        return default_from_dict(cls, data)
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]: These methods should not be needed anymore. Serialization should be covered by default serialization of components (see https://docs.haystack.deepset.ai/docs/serialization#default-serialization-behavior).

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    ) -> None:
+        """
+        Initialize the FirecrawlWebSearch component.
+
+        :param api_key:
+            API key for Firecrawl.
+            Defaults to the `FIRECRAWL_API_KEY` environment variable.
+        :param top_k:
+            Maximum number of documents to return.
+            Defaults to 10.
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]: Let's add here that this can be overridden by the `"limit"` parameter in `search_params`.

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            )
+        except Exception as error:
+            logger.exception(f"Failed to search for query '{query}': {error}")
+            return {"documents": [], "links": []}
+
+        documents, links = self._parse_search_response(search_response)
+
+        if self.top_k is not None:
+            documents = documents[: self.top_k]
+            links = links[: self.top_k]
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]: I think we can remove this, Firecrawl should return only `"limit"` search results.

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]: I think we can remove this, Firecrawl should return only `"limit"` search results.

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+        return {"documents": documents, "links": links}
+
+    @staticmethod
+    def _parse_search_response(search_response: Any) -> tuple[list[Document], list[str]]:
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]: Let's not use `Any` type here, this should be of type [`SearchData`](https://github.com/firecrawl/firecrawl/blob/5a5d4c86433e6d32682946d58d1f911efc423b8d/apps/python-sdk/firecrawl/v2/types.py#L1102).

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        """
+        Convert a Firecrawl search response to Haystack Documents and links.
+
+        :param search_response: Firecrawl search response object.
+        :returns: Tuple of (documents, links).
+        """
+        documents: list[Document] = []
+        links: list[str] = []
+
+        web_results = getattr(search_response, "web", None) or []
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]: Using the `SearchData` type for search_response, the following should work:

```suggestion
        web_results = search_response.web or []
```

File: integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+            if hasattr(result, "markdown") and result.markdown:
+                content = result.markdown
+                metadata = result.metadata_dict if hasattr(result, "metadata_dict") else {}
+                url = metadata.get("url", getattr(result, "url", ""))
+                title = metadata.get("title", "")
+            else:
+                url = getattr(result, "url", "") or ""
+                title = getattr(result, "title", "") or ""
+                content = getattr(result, "description", "") or ""
```
Comment (bogdankostic) [inline integrations/firecrawl/src/haystack_integrations/components/websearch/firecrawl/firecrawl_websearch.py]:
```suggestion
                url = getattr(result, "url", "")
                title = getattr(result, "title", "")
                content = getattr(result, "description", "")
```

**Review by bogdankostic — CHANGES_REQUESTED** [review]
Thanks for the PR @saakshigupta2002! I left a few comments on how to further improve it :)
